src/runtime/analytics-service.js

```diff
@@ -0,0 +1,201 @@
+const iframeAttributes = {
+  'visibility': 'hidden',
+  'opacity': 0,
+};
+
+export class AnalyticsService {
+  /**
+   * @param {!../model/doc.Doc} doc
+   * @param {!web-activities/activity-ports.ActivityPorts} activityPorts
+   * @param {!Object<string, ?>=} args
+   */
+  constructor(doc, activityPorts, args) {
+
+    /** @private @const {!Doc} */
+    this.doc_ = doc;
+
+    /** @private @const {!web-activities/activity-ports.ActivityPorts} */
+    this.activityPorts_ = activityPorts;
+
+    /** @private @const {!HTMLIFrameElement} */
+    this.iframe_ =
+        /** @type {!HTMLIFrameElement} */ (createElement(
+            this.doc_.getWin().document, 'iframe', iframeAttributes));
+
+    /** @private @const {string} */
+    this.src_ = feUrl('/serviceiframe');
+
+    this.args_ = feArgs({
+      publicationId: args['publicationId'],
+      // TODO(sohanirao): Add analytics context here
+    });
+    /** @private {?web-activities/activity-ports.ActivityIframePort} */
+    this.port_ = null;
+
+    /**
+     * @private
+     * {?function<!web-activities/activity-ports.ActivityIframePort|!Promise>}
+     */
+    this.portResolver_ = null;
+
+    /**
+     * @private @const
+     * {!Promise<!web-activities/activity-ports.ActivityIframePort>}
+     */
+    this.portPromise_ = new Promise(resolve => {
+      this.portResolver_ = resolve;
+    });
+
+    /**
+     * @private @constant
+     * {!FriendlyIframe}
+     */
+    this.friendlyIframe_ = new FriendlyIframe(this.doc_.getWin().document,
+        {'visibility': 'hidden'});
+
+    this.container_ = null;
+  }
+
+  /**
+   * @return {!HTMLIFrameElement}
+   */
+  getElement() {
+    return this.iframe_;
+  }
+
+  /**
+   * Build the iframe with the styling after iframe is loaded.
+   * @private
+   */
+  buildIframe_() {
+    const iframeBody = this.friendlyIframe_.getBody();
+    const iframeDoc =
+        /** @type {!HTMLDocument} */ (this.friendlyIframe_.getDocument());
+    // Container for all dynamic content, including 3P iframe.
+    const container = createElement(iframeDoc, 'swg-container',
+        {'visibility': 'hidden'});
+    iframeBody.appendChild(container);
+    return container;
+  }
+
+  getIframe() {
+    return this.friendlyIframe_;
+  }
+
+  open_() {
+    // Attach.
+    this.doc_.getBody().appendChild(this.friendlyIframe_.getElement());  // Fires onload.
+    return this.friendlyIframe_.whenReady().then(() => {
+      this.container_ = this.buildIframe_();
+    });
+  }
+
+  /**
+   * @return {!Promise}
+   */
+  init_() {
+    this.container_.appendChild(this.getElement());
+    return this.activityPorts_.openIframe(this.iframe_, this.src_,
+        this.args_).then(port => this.onOpenIframeResponse_(port));
+  }
+
+  /**
+   * @param {!web-activities/activity-ports.ActivityIframePort} port
+   * @return {!Promise}
+   */
+  onOpenIframeResponse_(port) {
+    this.port_ = port;
+    this.portResolver_(port);
+    return this.port_.whenReady();
```

Comment:
We don't need to wait for port readiness: it's specifically a UI signal. It basically means "ready to be displayed". Probably a slightly unfortunate name. But otherwise - as soon as we are connected - we are all good to go.